File: windows-client/src/app/services/game-finder/game-finder.service.ts
import { Injectable } from '@angular/core';
import { CommunicationSocketService } from '@app/services/communication-socket/communication-socket.service';
import { GameRoom } from '@common/classes/game-room';
import { GameFinderEvents } from '@common/enums/game-finder.gateway.variables';
import { GameMode } from '@common/game-mode';
import { Observable, Subject } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class GameFinderService {
    gameExists$ = new Subject<boolean>();
    gameMode: string;
    constructor(private readonly socketService: CommunicationSocketService) {}

    getGames(gameName = undefined as unknown as string): Observable<GameRoom[]> {
        return new Observable((observer) => {
            this.socketService.on(GameFinderEvents.Games, (data: { games: GameRoom[]; gameName: string; gameMode: string }) => {
                if (
                    (this.gameMode === GameMode.limitedTimeMode && data.gameMode === GameMode.limitedTimeMode) ||
                    (this.gameMode === GameMode.classicMode && data.gameName === gameName && data.gameMode === GameMode.classicMode)
                ) {
                    observer.next(data.games);
                }
            });
            this.socketService.send(GameFinderEvents.GetGames, { gameMode: this.gameMode, gameName });
        });
    }
}

File: android-client/lib/services/game_finder_service.dart
import 'package:android_client/classes/game_room.dart';
import 'package:android_client/services/socket_service.dart';
import 'package:rxdart/rxdart.dart';

class GameFinderService {
  final SocketService _socketService = SocketService();
  late String gameMode;
  // StreamController<bool> gameExists$ = StreamController<bool>();

  GameFinderService(this.gameMode);

  
  BehaviorSubject<List<GameRoom>> getGames(String? gameName) {
    final controller = BehaviorSubject<List<GameRoom>>();
    
    _socketService.on('games', (dynamic data) {
      if ((gameMode == 'mode Temps Limité' && data['gameMode'] == 'mode Temps Limité') ||
        (gameMode == 'mode Classique' && data['gameName'] == gameName && data['gameMode'] ==  'mode Classique')) {
        if (data['games'].length > 0) {
        controller.add((data['games'] as List).map((gameData) => GameRoom.fromJson(gameData)).toList());
        } else {
          controller.add([]);
        }
      }
    });
    
    _socketService.send('getGames', {'gameMode': gameMode, 'gameName': gameName});
    
    return controller;
  }

}
